modules/internal/ChapelLocale.chpl

```diff
@@ -90,20 +90,142 @@ module ChapelLocale {
     return (subloc != c_sublocid_none
             && subloc != c_sublocid_any
             && subloc != c_sublocid_all);
+  
+  /*
+    regular: Has a concrete BaseLocale instance
+    nilLocale: The _instance is set to nil. Used during setup. Also, as a
+               sentinel value in locale tree operations
+    dummy: Used during initialization for `here` before it is properly setup
+    default: Used to store the default locale instance. Initially set to nil,
+             then "fixed" by LocalesArray to Locales[0]
+    any: Placeholder to represent the notion of "anywhere"
+   */
+  pragma "no doc"
+  enum localeKind { regular, any, nilLocale, dummy, default };
+
+  const nilLocale = new locale(localeKind.nilLocale);
+  var defaultLocale = new locale(localeKind.default);
+
+  pragma "always RVF"
+  record _locale {
+    var _instance: unmanaged BaseLocale?;
```

Comment:
Is the only complication for making this use a non-nilable class the fact that the `nilLocale` value stores `nil` here?  What would be the impact if the `nilLocale` stored a non-nil class instance (on locale 0, say), to remove the need for nil-checking and the potential for halts if it fails?  (e.g., is there code that could get into trouble by calling `nilLocale._value`?)  Or by storing a `kind` field?

I also find myself wondering whether `_locale` should store an `owned` field for the original locales so that they can delete their instances at the end of the program?  Or does something else do that cleanup?  (I was just noting that the deinit()s were empty).

Reply:
Hmm, I haven't tried the first part. I was (and am) a bit worried about the initialization order of things. Having this field default to nil allows us to create `locale` variables without triggering any initializer that may have dependencies on some other things (defaultDist for example) that are not initialized at this point. I am a bit wary of playing with these as they are prone to subtle errors. That being said, I can try. However, I am curious about the motivation for that, too.

With the current version of this branch, deinitializers are used appropriately and locale instances are cleaned up properly. (Tested with `valgrind --leak-check=yes` manually). I started implementing an `isOwned` field, but couldn't clear up the dependencies in a reasonable amount of time. I have it in a separate commit and can revisit next week.

Reply:
The motivation is just that if you were to ask me "Does every locale value have a class associated with it?" my quick reaction is "I expect and hope so" and then "does a locale record change or delete its related class during its lifetime?" my sense is "I hope not."  Combining that with a sense that it's most attractive to use non-nilable whenever possible, my guess would've been that locales would be a natural candidate for non-nilable.  So then I was trying to guess why they weren't.

Reply:
I played around a bit with this. Based on that and our offline discussion, I am inclined to postpone trying this. It is not straightforward as it may seem because of dependencies and the order of initialization between multiple locale-related modules.

To capture some of my hesitation that I already mentioned offline: I understand that the code for the locale implementation might be cleaner if `_instance` was non-nilable. However, as this branch currently stands, the user will never get a `locale` whose instance is `nil`. From a more philosophical standpoint. it makes sense to me that a `BaseLocale` instance can be nil because of all the intricacies that happen until we have a full-blown `Locales` array.

Reply:
That's fine with me, thanks for looking into it.